src/app-layer-parser.c

```diff
@@ -1234,12 +1238,35 @@ int AppLayerParserParse(ThreadVars *tv, AppLayerParserThreadCtx *alp_tctx, Flow
     /* invoke the recursive parser, but only on data. We may get empty msgs on EOF */
     if (input_len > 0 || (flags & STREAM_EOF)) {
         /* invoke the parser */
-        if (p->Parser[(flags & STREAM_TOSERVER) ? 0 : 1](f, alstate, pstate,
+        AppLayerReturn res = p->Parser[direction](f, alstate, pstate,
                 input, input_len,
                 alp_tctx->alproto_local_storage[f->protomap][alproto],
-                flags) < 0)
+                flags);
+        if (res.status < 0)
         {
             goto error;
+        } else if (res.status > 0) {
+            TcpSession *ssn = f->protoctx;
+            if (ssn != NULL) {
+                SCLogDebug("direction %d/%s", direction, (flags & STREAM_TOSERVER) ? "toserver":"toclient");
+                assert(res.consumed <= input_len);
```

Comment:
BUG_ON?

Reply:
Yeah. This should probably become a DEBUG_VALIDATE_BUG_ON later, but will leave it as a BUG_ON in the dev tree for a while